crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -435,26 +435,59 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> DatabaseProvider<'this, TX> {
             self.get_or_take::<tables::TxSenders, TAKE>(first_transaction..=last_transaction)?;
 
         // Recover senders manually if not found in db
-        let senders_len = senders.len();
-        let transactions_len = transactions.len();
-        let missing_senders = transactions_len - senders_len;
-        let mut senders_recovered: Vec<(u64, H160)> = (first_transaction..
-            first_transaction + missing_senders as u64)
-            .zip(
-                TransactionSigned::recover_signers(
-                    transactions.iter().take(missing_senders).map(|(_, tx)| tx).collect::<Vec<_>>(),
-                    missing_senders,
-                )
-                .ok_or(BlockExecutionError::Validation(
-                    BlockValidationError::SenderRecoveryError,
-                ))?,
+        if senders.len() != transactions.len() {
+            // Find all missing senders, their corresponding tx numbers and indexes to the original
+            // `senders` vector at which the recovered senders will be inserted
+            let mut missing_senders = Vec::with_capacity(transactions.len() - senders.len());
+            {
+                let mut senders = senders.iter().peekable();
+
+                // `transactions` contain all entries. `senders` contain _some_ of the senders for
+                // these transactions. Both are sorted and indexed by `TxNumber`.
+                //
+                // The general idea is to iterate on both `transactions` and `senders`, and advance
+                // the `senders` iteration only if it matches the current `transactions` entry's
+                // `TxNumber`. Otherwise, add the transaction to the list of missing senders.
+                for (i, (tx_number, transaction)) in transactions.iter().enumerate() {
+                    if let Some((sender_tx_number, _)) = senders.peek() {
+                        if sender_tx_number == tx_number {
+                            // If current sender's `TxNumber` matches current transaction's
+                            // `TxNumber`, advance the senders iterator.
+                            senders.next();
+                        } else {
+                            // If current sender's `TxNumber` doesn't match current transaction's
+                            // `TxNumber`, add it to missing senders.
+                            missing_senders.push((i, tx_number, transaction));
+                        }
+                    } else {
+                        // If there's no more senders left, but we're still iterating over
+                        // transactions, add them to missing senders
+                        missing_senders.push((i, tx_number, transaction));
+                    }
```

Comment:
```suggestion
                    if lsenders.peek().map_or(false, |(sender_tx_number, _)| sender_tx_number == tx_number) {
                        // If current sender's `TxNumber` matches current transaction's
                        // `TxNumber`, advance the senders iterator.
                        senders.next();
                    } else {
                        // If current sender's `TxNumber` doesn't match current transaction's or
                        // there're no more senders left, but we're still iterating over transactions,
                        // add them to missing senders
                        missing_senders.push((i, tx_number, transaction));
                    }
```

Reply:
I made it even more verbose, because I personally often get confused by `map_or(false)` or `map_or(true)`